py/desisim/bal.py

```diff
@@ -0,0 +1,108 @@
+"""
+desisim.bal
+===========
+
+Functions and methods for inserting BALs into QSO spectra.
+
+"""
+from __future__ import division, print_function
+
+import numpy as np
+from astropy.io import fits 
+import random 
+
+def readbaltemplates(file):
+    """ Read in the BAL templates fits file
+    Args: 
+	file: name of fits file with BAL templates
+    Returns: 
+	wave_bal_template (ndarray): 1-D wavelength array in Ang for 
+	  all templates in rest frame 
+	bal_templates (ndarray): 2-d array with all BAL templates 
+    """ 
+
+    hdu = fits.open(file)
+    bal_templates = hdu[1].data['TEMP'] 
+    nlam = len(hdu[1].data['TEMP'][0])	# number of wavelength elements
+    lam1 = hdu[1].header['CRVAL1']
+    dlam = hdu[1].header['CDELT1']
+    lam2 = lam1 + dlam*nlam
+    wave_bal_template = np.arange(lam1, lam2, dlam)
+    return wave_bal_template, bal_templates 
+
+
+def isbal(balprob): 
+    """ Randomly determine if a QSO is a BAL 
+    Args: 
+	balprob (float) : probability that a QSO is a BAL 
+	Expect about 12% based on DR12
+    Returns: 
+	booleal: True if a BAL, otherwise False
+    """ 
+
+    # Make sure probability is on the interval [0.,1.]
+    if balprob > 1.:
+        balprob = 1. 
+    if balprob < 0.: 
+        balprob = 0.
+
+    if np.random.random() < balprob:
+        return True
+    return False
+
+
+def getbaltemplate(qsowave, qsoredshift, balwave, baltemplates):
+    """ Get BAL template for a single QSO. 
```

Comment:
I wonder if a lot of the functionality here can be done *once* by instantiating a class, so that when we generate many templates we don't have to redo many of these calculations (e.g., extending the template in wavelength and adding in the CIV template).

For example, the [SUPERSTAR](https://github.com/desihub/desisim/blob/master/py/desisim/templates.py#L1014) class may be a useful model.

Reply:
I've changed the templates so they no longer need to be extended in wavelength each time.